src/fifelse.c

```diff
@@ -0,0 +1,338 @@
+      for(i = 0; i < len0; i++)
+      {
+        switch(pl[i])
+        {
+        case 0  : int_pres[i] = int_pb[i];  break;
+        case 1  : int_pres[i] = int_pa[i];  break;
+        default : int_pres[i] = NA_LOGICAL; break;
+        }
+      }
+      break;
+      
+    case 1:
+      for(i = 0; i < len0; i++)
+      {
+        switch(pl[i])
+        {
+        case 0  : int_pres[i] = int_pb[i];  break;
+        case 1  : int_pres[i] = int_pa[0];  break;
+        default : int_pres[i] = NA_LOGICAL; break;
+        }
+      }
+      break;
+      
+    case 2:
+      for(i = 0; i < len0; i++)
+      {
+        switch(pl[i])
+        {
+        case 0  : int_pres[i] = int_pb[0];  break;
+        case 1  : int_pres[i] = int_pa[i];  break;
+        default : int_pres[i] = NA_LOGICAL; break;
+        }
+      }
+      break;
+      
+    case 3:
+      for(i = 0; i < len0; i++)
+      {
+        switch(pl[i])
+        {
+        case 0  : int_pres[i] = int_pb[0];  break;
+        case 1  : int_pres[i] = int_pa[0];  break;
+        default : int_pres[i] = NA_LOGICAL; break;
+        }
+      }
+      break;
+    }
+    break;
+    
+    /*
+     This part is for dealing with a (yes) and b (no)
+     in case they are both of type STRSXP (logical)
+     */
+  case STRSXP :
+    result = PROTECT(allocVector(STRSXP, len0));
+    stack_size++;
+    switch(adj)
+    {
+    case 0:
+      for(i = 0; i < len0; i++)
+      {
+        switch(pl[i])
+        {
+        case 0  : SET_STRING_ELT(result, i, STRING_ELT(b, i));  break;
+        case 1  : SET_STRING_ELT(result, i, STRING_ELT(a, i));  break;
+        default : SET_STRING_ELT(result, i, NA_STRING); break;
+        }
+      }
+      break;
+      
+    case 1:
+      for(i = 0; i < len0; i++)
+      {
+        switch(pl[i])
+        {
+        case 0  : SET_STRING_ELT(result, i, STRING_ELT(b, i));  break;
+        case 1  : SET_STRING_ELT(result, i, STRING_ELT(a, 0));  break;
+        default : SET_STRING_ELT(result, i, NA_STRING); break;
+        }
+      }
+      break;
+      
+    case 2:
+      for(i = 0; i < len0; i++)
+      {
+        switch(pl[i])
+        {
+        case 0  : SET_STRING_ELT(result, i, STRING_ELT(b, 0));  break;
+        case 1  : SET_STRING_ELT(result, i, STRING_ELT(a, i));  break;
+        default : SET_STRING_ELT(result, i, NA_STRING); break;
+        }
+      }
+      break;
+      
+    case 3:
+      for(i = 0; i < len0; i++)
+      {
+        switch(pl[i])
+        {
+        case 0  : SET_STRING_ELT(result, i, STRING_ELT(b, 0));  break;
+        case 1  : SET_STRING_ELT(result, i, STRING_ELT(a, 0));  break;
+        default : SET_STRING_ELT(result, i, NA_STRING); break;
+        }
+      }
+      break;
+    }
+    break;
+    
+  default: error("Type %s is not supported.",type2char(ta)); break;
+  }
+  
+  // Check if class type is Date and adjust
+  if(!isNull(class_type))
+    if( CHAR(STRING_ELT(class_type,0)) == CHAR(STRING_ELT(Rf_mkString("Date"),0)) )
```

Comment:
We have such character strings predefined in init.c, those should be preferred over mkString